internal/manager/controllerdef.go

```diff
@@ -151,110 +205,74 @@ func setupControllers(logger logr.Logger, mgr manager.Manager, proxy proxy.Proxy
-	if err == nil {
-		controllers = append(controllers, ingressControllers[negotiatedIngressAPI])
-	} else {
-		logger.Info(`no Ingress controllers enabled or no suitable Ingress version found.
-		Disabling Ingress controller`)
-	}
+	return controllers, nil
+}
 
-	if c.KongClusterPluginEnabled == util.EnablementStatusEnabled {
-		kongClusterPluginGVR := schema.GroupVersionResource{
-			Group:    konghqcomv1.SchemeGroupVersion.Group,
-			Version:  konghqcomv1.SchemeGroupVersion.Version,
-			Resource: "kongclusterplugins",
-		}
+// crdExistsChecker verifies whether the resource type defined by GVR is supported by the k8s apiserver.
+type crdExistsChecker struct {
+	GVR schema.GroupVersionResource
+}
 
-		if ctrlutils.CRDExists(mgr.GetClient(), kongClusterPluginGVR) {
-			logger.Info("kongclusterplugins.configuration.konghq.com v1beta1 CRD available on cluster.")
-			controller := ControllerDef{
-				IsEnabled: &c.KongClusterPluginEnabled,
-				Controller: &configuration.KongV1KongClusterPluginReconciler{
-					Client:           mgr.GetClient(),
-					Log:              ctrl.Log.WithName("controllers").WithName("KongClusterPlugin"),
-					Scheme:           mgr.GetScheme(),
-					Proxy:            proxy,
-					IngressClassName: c.IngressClassName,
-				},
-			}
-			controllers = append(controllers, controller)
-		} else {
-			message := fmt.Sprintf("%s CRD not available on the cluster", kongClusterPluginGVR.String())
-			return nil, fmt.Errorf(message)
-		}
-	} else {
-		logger.Info(`kong cluster plugin is disabled.
-		Disabling KongClusterPlugin controller`)
-	}
-	if c.KnativeIngressEnabled == util.EnablementStatusEnabled {
-		knativeGVR := schema.GroupVersionResource{
-			Group:    knativev1alpha1.SchemeGroupVersion.Group,
-			Version:  knativev1alpha1.SchemeGroupVersion.Version,
-			Resource: "ingresses",
-		}
+// CRDExists returns true iff the apiserver supports the specified group/version/resource.
+func (c crdExistsChecker) CRDExists(r client.Client) bool {
+	return ctrlutils.CRDExists(r, c.GVR)
+}
 
-		if ctrlutils.CRDExists(mgr.GetClient(), knativeGVR) {
-			logger.Info("ingresses.networking.internal.knative.dev v1alpha1 CRD available on cluster.")
-			controller := ControllerDef{
-				IsEnabled: &c.KnativeIngressEnabled,
-				Controller: &configuration.Knativev1alpha1IngressReconciler{
-					Client:           mgr.GetClient(),
-					Log:              ctrl.Log.WithName("controllers").WithName("Ingress").WithName("KnativeV1Alpha1"),
-					Scheme:           mgr.GetScheme(),
-					Proxy:            proxy,
-					IngressClassName: c.IngressClassName,
-				},
-			}
-			controllers = append(controllers, controller)
-		} else {
-			message := fmt.Sprintf("%s CRD not available on the cluster", knativeGVR.String())
-			return nil, fmt.Errorf(message)
-		}
-	} else {
-		logger.Info(`knative v1alpha1 ingress is disabled.
-		Disabling Knative controller`)
-	}
+// ingressControllerStrategy picks the best Ingress API supported by k8s apiserver.
```

Comment:
do you refer best the latest ? most stable ?

Reply:
The latest which is not disabled by flags.

In practice, we'll retire everything but v1 in a few months - see #1666